File: src/SolitaireStack.java
import java.util.Stack;

public abstract class SolitaireStack {
    private Stack<Card> cards;

    public SolitaireStack(Stack<Card> cards) {
        this.cards = cards;
    }

    public Stack<Card> getCards() {
        return cards;
    }

    public boolean isEmpty(){
        return this.getCards().empty();
    }

    public abstract boolean canMove(SolitaireStack sourceStack);

    public void move(SolitaireStack sourceStack){
        Card newCard = sourceStack.top();
        Card topCard = this.isEmpty() ? null : this.top();

        this.add(sourceStack.getCards().pop());
        Global.logMove(this.moveMessage(topCard, newCard));
    }

    protected abstract String moveMessage(Card topCard, Card newCard);

    public void add(Card c){
        this.getCards().push(c);
    }

    public Card top(){
        return this.getCards().peek();
    }
}

File: solitaire_stack.py
from collections import deque

import file_manager


class SolitaireStack:
    def __init__(self):
        self.cards = deque()

    def __eq__(self, other):
        return self.cards == other.cards

    def __ne__(self, other):
        return self.cards != other.cards

    def __hash__(self):
        return hash(tuple(self.cards))

    def is_empty(self):
        return not self.cards

    def can_move(self, source_stack):
        pass

    def move(self, source_stack):
        new = source_stack.cards.pop()
        top = self.top()

        self.add(new)

        return self.move_message(top, new)

    def add(self, card):
        self.cards.append(card)

    def top(self):
        if self.cards:
            return self.cards[-1]
        return None

    def move_message(self, top, new):
        pass
